stockroom/diff.py

```diff
@@ -0,0 +1,126 @@
+from typing import NamedTuple
+
+from hangar import Repository
+from rich.jupyter import print
+from rich.table import Table
+from rich.segment import Segment
+from rich.console import Console
+
+from .parser import tagkey, SEP
+
+class ParsedDiff(NamedTuple):
+    tagName: str
+    ref_sample: str = ''
+    dest_samples: list = list()
+
+class TagDiff:
+    """
+    The Diff object for tags
+    """
+
+    def __init__(self, ref_commit: str = None, dest_commit: str = None,
+                    root: str = None):
+        self.ref_commit = ref_commit
+        self.dest_commit = dest_commit
+        self.repo = Repository(root)
+        self.all_keys = set()
+
+        self._added: dict = dict()
+        self._deleted: dict = dict()
+        self._mutated: dict = dict()
+
+        self.parse_diff()
+
+    def gen_tables(self):
+    """
+    Generate the Rich Table using parsed data.
+    """
+
+        table = Table(title='Diff %s..%s'%(self.ref_commit[2:7], self.dest_commit[2:7]))
+
+        # Generate Columns
+        column_headings = ['Tags'] + [self.ref_commit, self.dest_commit]
+        for column in column_headings:
+            table.add_column(column[:10], no_wrap=True)
+
+        # Generate rows
+        for row in self.all_keys:
+            ref_row = ""
+            dest_row = ""
+
+            if row in self._added:
+                dest_row = "[green] %s [/green]"%(self._added[row])
+
+            elif row in self._mutated:
+                ref_row = "[red] %s [/red]"%(self._mutated[row][0])
+                dest_row = "[red] %s [/red]"%(self._mutated[row][1])
+
+            elif row in self._deleted:
+                ref_row = "[green] %s [/green]"%(self._deleted[row])
+
+            table.add_row(row, ref_row, dest_row)
+        return table
+
+    def parse_diff(self):
+        """
+        Parse the hangar.diff and store the changes in the TagDiff object
+        """
+
+        print('Accessing ref_commit')
```

Comment:
I don't think we should print these messages; do you have any specific reason to think would help the user?

Reply:
No I guess not now ... but maybe later when we have the capabilities to diff with multiple branches and staged? seemed like a good info but does clutter it.

Reply:
@hhsecond I think that for the purposes of Stockroom, it would be a good idea to implement some sort of logging which records the series of operations performed...

As we've discussed, managing branching/merging will be a bit of a coordination challenge (syncronizing branches between Git & Hangar), and i can imagine there are a few corner cases where having an audit log of 

```
accessed commit #foo
diffed commit #foo << #bar
checking out tag "vFoo.Bar.Baz"
created branch xyz
...
```

would be super helpful for debugging purposes. 

The hangar core can reconstruct the sequence of operations by querying it's internals directly, but since that isn't available to stockroom (It's not a public API, and never will be) it wouldn't be the worst idea to use a logging file handler to record stuff like this (even if it's not printed to the screen). 

Just my 2-cents worth of input :)